core/src/main/scala/dogs/Heap.scala

```diff
@@ -0,0 +1,207 @@
+ *
+ * It is important to note that we can, in fact, to create the Binary Heap in order O(n) from a `List` using the function
+ * `heapify`.
+ */
+sealed abstract class Heap[A] {
+
+  import Heap._
+
+  /**
+   * Internal representation of the min value to avoid deconstruction of `min: Option[A]` since min is heavily used.
+   */
+  private [dogs] val min: A
+
+  /**
+   * Returns min value on the heap.
+   */
+  def getMin: Option[A]
+
+  private [dogs] def left: Heap[A]
+
+  private [dogs] def right: Heap[A]
+
+  /**
+   * Returns the size of the heap.
+   */
+  def size: Int
+
+  /**
+   * Returns the height of the heap.
+   */
+  def height: Int
+
+  /**
+   * Verfies if the heap is empty.
+   */
+  def isEmpty: Boolean
+
+
+  /**
+   * Insert a new element into the heap.
+   * Order O(log n)
+   */
+  def add(x: A)(implicit order: Order[A]): Heap[A] =
+    if (isEmpty)
+      Heap(x, Leaf(), Leaf())
+    else if (left.size < scala.math.pow(2, left.height) - 1)
+      bubbleUp(min, left.add(x), right)
+    else if (right.size < scala.math.pow(2, right.height) - 1)
+      bubbleUp(min, left, right.add(x))
+    else if (right.height < left.height)
+      bubbleUp(min, left, right.add(x))
+    else
+      bubbleUp(min, left.add(x), right)
+
+  /**
+   * Build a heap using a list.
+   * Order O(n)
+   */
+  def heapify(a: List[A])(implicit order: Order[A]): Heap[A] = {
+    def loop(i: Int, xs: scala.List[A]): Heap[A] =
+      if (i < xs.length) {
+        bubbleDown(xs(i), loop(2 * i + 1, xs), loop(2 * i + 2, xs))
+      }
+      else {
+        Leaf()
+      }
+
+    loop(0, a.toScalaList)
+  }
+
+  /**
+   * Remove the min element from the heap (the root).
+   * Order O(log n)
+   */
+  def remove()(implicit order: Order[A]): Heap[A] = this match {
+    case Leaf()                 =>  Leaf()
+    case Branch(_, l, r, _, _)  =>  bubbleRootDown(mergeChildren(l, r))
+  }
+
+  /**
+   * Returns a sorted list of the elements within the heap.
+   */
+  def toList()(implicit order: Order[A]): List[A] = this match {
+    case Leaf()                       =>  El[A]
+    case Branch(m, l, r, _, _)        =>  Nel(m, remove.toList)
+  }
+
+  /**
+   * Alias for add
+   */
+  def +(x: A)(implicit order: Order[A]): Heap[A] = add(x)
+
+  /**
+   * Alias for remove
+   */
+  def --(implicit order: Order[A]): Heap[A] = remove()
+
+}
+
+object Heap {
+
+  def empty[A]: Heap[A] = Leaf()
+
+  def apply[A](x: A): Heap[A] = Branch(x, empty, empty, 1, 1)
+
+  def apply[A](x: A, l: Heap[A], r: Heap[A]): Heap[A] =
+    Branch(x, l, r, l.size + r.size + 1, scala.math.max(l.height, r.height) + 1)
+
+  private [dogs] case class Branch[A](min: A, left: Heap[A], right: Heap[A], size: Int, height: Int) extends Heap[A] {
+    override def isEmpty: Boolean = false
+
+    override def getMin: Option[A] = Some(min)
+  }
+
+  private [dogs] case object Leaf extends Heap[Option[Nothing]] {
```

Comment:
min in Leaf is a nothing. How do I return Nothing? 

val min: Nothing = ????  <- what to put in here